pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java

```diff
@@ -120,14 +108,17 @@ public String toExplainString() {
   protected TransferableBlock getNextBlock() {
     try {
       cumulateAggregationBlocks();
-      return new TransferableBlock(toResultBlock());
+      return toResultBlock();
     } catch (Exception e) {
       return TransferableBlockUtils.getErrorTransferableBlock(e);
     }
   }
 
-  private BaseDataBlock toResultBlock()
+  private TransferableBlock toResultBlock()
       throws IOException {
+    if (_upstreamErrorBlock != null) {
```

Comment:
Is my following understanding correct regarding error handling ?

Say stage 0 ,1, 2 from root to leaf. 

Regardless of which operator in a given stage is the main origin of error, the broadcast of errors to downstream operators / stages is always done by the sender / root operator in the stage that caught error ?

For example, stage 2 is running on a node N and has 3 operators O1 -> O2 -> O3. If O3 catches error, it is still going to the return a block to O2 and O2 to O1 that indicates error, but O1 (which should be the sender) will create a TransferableBlock for error and send to all receivers in the downstream stage ?

Does that mean when stages are setup, regardless of the exchange type between two stages, each sender in the  upstream stage has to be aware of each receiver in the downstream stage ?

Reply:
everything is correct, except for the transferableBlock with error 
it will only send to the receiver that are originally expect from this sender. not to everyone.

Reply:
This is tricky, so long-story short. the mailbox send/receive is only special because of the "random distribution" exchange type. 
without it. the connection will be fully connected (for data block)
- in either broadcast or hash distribution, all sender servers will send data to all receiver servers
- in singleton distribution, the list of sender and receiver will only do self-connection.
- where as in random distribution, we send data to one of the receiver servers

however, for metadata block. EXACTLY BECAUSE the behavior of random distribution, we need to broadcast the end-of-stream info to every receiver since they might've still expect a data block to come if we dont do so. 

so in summary
- error blocks are broadcast to all receivers.
- except for singleton, since they established the 1-to-1 relation at the beginning.